datadog/fwprovider/resource_datadog_restriction_policy.go

```diff
@@ -0,0 +1,240 @@
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	resourceId := state.ResourceId.ValueString()
+	body, diags := r.buildRestrictionPolicyRequestBody(ctx, &state)
+	response.Diagnostics.Append(diags...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	resp, _, err := r.API.UpdateRestrictionPolicy(r.Auth, resourceId, *body)
+	if err != nil {
+		response.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error retrieving RestrictionPolicy"))
+		return
+	}
+	if err := utils.CheckForUnparsed(resp); err != nil {
+		response.Diagnostics.AddError("response contains unparsedObject", err.Error())
+		return
+	}
+	r.updateState(ctx, &state, &resp)
+
+	// Save data into Terraform state
+	response.Diagnostics.Append(response.State.Set(ctx, &state)...)
+}
+
+func (r *RestrictionPolicyResource) Update(ctx context.Context, request resource.UpdateRequest, response *resource.UpdateResponse) {
+	var state RestrictionPolicyModel
+	response.Diagnostics.Append(request.Plan.Get(ctx, &state)...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	resourceId := state.ResourceId.ValueString()
+	body, diags := r.buildRestrictionPolicyRequestBody(ctx, &state)
+	response.Diagnostics.Append(diags...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	resp, _, err := r.API.UpdateRestrictionPolicy(r.Auth, resourceId, *body)
+	if err != nil {
+		response.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error retrieving RestrictionPolicy"))
+		return
+	}
+	if err := utils.CheckForUnparsed(resp); err != nil {
+		response.Diagnostics.AddError("response contains unparsedObject", err.Error())
+		return
+	}
+	r.updateState(ctx, &state, &resp)
+
+	// Save data into Terraform state
+	response.Diagnostics.Append(response.State.Set(ctx, &state)...)
+}
+
+func (r *RestrictionPolicyResource) Delete(ctx context.Context, request resource.DeleteRequest, response *resource.DeleteResponse) {
+	var state RestrictionPolicyModel
+	response.Diagnostics.Append(request.State.Get(ctx, &state)...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	id := state.ID.ValueString()
+	httpResp, err := r.API.DeleteRestrictionPolicy(r.Auth, id)
+	if err != nil {
+		if httpResp != nil && httpResp.StatusCode == 404 {
+			return
+		}
+		response.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error deleting restriction_policy"))
+		return
+	}
+}
+
+func (r *RestrictionPolicyResource) updateState(ctx context.Context, state *RestrictionPolicyModel, resp *datadogV2.RestrictionPolicyResponse) {
+	state.ID = types.StringValue(resp.Data.GetId())
+	state.ResourceId = types.StringValue(resp.Data.GetId())
+	data := resp.GetData()
+	attributes := data.GetAttributes()
+
+	if bindings, ok := attributes.GetBindingsOk(); ok && len(*bindings) > 0 {
+		state.Bindings = []*BindingsModel{}
+		for _, bindingsDd := range *bindings {
+			bindingsTfItem := BindingsModel{}
+			if principals, ok := bindingsDd.GetPrincipalsOk(); ok && len(*principals) > 0 {
```

Comment:
just curious, if the principals is empty, shall we proceed this logic?

Reply:
I mean adding bindingsTfItem to the Bindings

Reply:
We have validation in place to prevent empty principal lists.
Example call:
```
resource "datadog_restriction_policy" "foo" {
  resource_id = "notebook:id"
  bindings {
    principals = []
    relation = "editor"
  }
}
```

Result:
```
Error: error retrieving RestrictionPolicy 
with datadog_restriction_policy.foo,
on main.tf line 14, in resource "datadog_restriction_policy" "foo":
14: resource "datadog_restriction_policy" "foo" {

400 Bad Request: {"errors":["Invalid request: principals list cannot be empty; found empty list for relation 'editor'"]}
```

Reply:
I see